x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/geo/GeoShape.java

```diff
@@ -63,6 +74,88 @@ public Geometry toGeometry() {
         return shapeBuilder.buildGeometry();
     }
 
+    public Point firstPoint() {
+        return shapeBuilder.buildGeometry().visit(new GeometryVisitor<Point>() {
+            @Override
+            public Point visit(Circle circle) {
+                return new Point(circle.getLat(), circle.getLon(), circle.hasAlt() ? circle.getAlt() : Double.NaN);
+            }
+
+            @Override
+            public Point visit(GeometryCollection<?> collection) {
+                if (collection.size() > 0) {
+                    return collection.get(0).visit(this);
+                }
+                return null;
+            }
+
+            @Override
+            public Point visit(Line line) {
+                if (line.length() > 0) {
+                    return new Point(line.getLat(0), line.getLon(0), line.hasAlt() ? line.getAlt(0) :  Double.NaN);
+                }
+                return null;
+            }
+
+            @Override
+            public Point visit(LinearRing ring) {
+                return visit((Line) ring);
+            }
+
+            @Override
+            public Point visit(MultiLine multiLine) {
+                return visit((GeometryCollection<?>) multiLine);
```

Comment:
I think here and below the cast is unnecessary?

Reply:
Without this cast it will go into an infinite loop.